IGSIOCalibration/vtkIGSIOPivotCalibrationAlgo.cxx

```diff
@@ -160,45 +150,128 @@ igsioStatus vtkIGSIOPivotCalibrationAlgo::GetPivotPointPosition(double* pivotPoi
+
+//----------------------------------------------------------------------------
+igsioStatus vtkIGSIOPivotCalibrationAlgo::DoPivotCalibration(vtkIGSIOTransformRepository* aTransformRepository/* = NULL*/, bool autoOrient)
+{
+  if (this->MarkerToReferenceTransformMatrixBuckets.empty())
+  {
+    this->SetErrorCode(CALIBRATION_NOT_ENOUGH_POINTS);
+    return IGSIO_FAIL;
+  }
+
+  if (this->GetMaximumToolOrientationDifferenceDeg() < this->MinimumOrientationDifferenceDeg)
+  {
+    this->SetErrorCode(CALIBRATION_NOT_ENOUGH_VARIATION);
+    return IGSIO_FAIL;
+  }
+
+  if (!this->PivotPointToMarkerTransformMatrix)
+  {
+    vtkNew<vtkMatrix4x4> pivotPointToMarkerTransformMatrix;
+    this->SetPivotPointToMarkerTransformMatrix(pivotPointToMarkerTransformMatrix);
+  }
+
+  double pivotPoint_Marker[4] = { 0.0, 0.0, 0.0, 1.0 };
+  double pivotPoint_Reference[4] = { 0.0, 0.0, 0.0, 1.0 };
+  std::vector<vtkMatrix4x4*> markerToTransformMatrixArray = this->GetAllMarkerToReferenceMatrices();
+  igsioStatus status = this->DoPivotCalibrationInternal(&markerToTransformMatrixArray, autoOrient, &this->OutlierIndices, pivotPoint_Marker, pivotPoint_Reference, this->PivotPointToMarkerTransformMatrix);
+  if (status == IGSIO_SUCCESS)
+  {
+    this->SetErrorCode(CALIBRATION_SUCCESS);
+  }
+  else
+  {
+    this->SetErrorCode(CALIBRATION_FAIL);
+  }
+
+  this->PivotPointPosition_Reference[0] = pivotPoint_Reference[0];
+  this->PivotPointPosition_Reference[1] = pivotPoint_Reference[1];
+  this->PivotPointPosition_Reference[2] = pivotPoint_Reference[2];
+
+  this->ComputePivotCalibrationError();
+
+  // Save result
+  if (aTransformRepository)
+  {
+    igsioTransformName pivotPointToMarkerTransformName(this->ObjectPivotPointCoordinateFrame, this->ObjectMarkerCoordinateFrame);
+    aTransformRepository->SetTransform(pivotPointToMarkerTransformName, this->PivotPointToMarkerTransformMatrix);
+    aTransformRepository->SetTransformPersistent(pivotPointToMarkerTransformName, true);
+    aTransformRepository->SetTransformDate(pivotPointToMarkerTransformName, vtkIGSIOAccurateTimer::GetInstance()->GetDateAndTimeString().c_str());
+    aTransformRepository->SetTransformError(pivotPointToMarkerTransformName, this->PivotCalibrationErrorMm);
+  }
+  else
+  {
+    LOG_DEBUG("Transform repository object is NULL, cannot save results into it");
+  }
+
+  return status;
+}
+
+//----------------------------------------------------------------------------
+igsioStatus vtkIGSIOPivotCalibrationAlgo::DoPivotCalibrationInternal(const std::vector<vtkMatrix4x4*>* markerToTransformMatrixArray, bool autoOrient, std::set<unsigned int>* outlierIndices, double pivotPoint_Marker[4], double pivotPoint_Reference[4], vtkMatrix4x4* pivotPointToMarkerTransformMatrix)
 {
-  if (this->MarkerToReferenceTransformMatrixArray.empty())
+  if (!markerToTransformMatrixArray || markerToTransformMatrixArray->size() == 0)
   {
-    LOG_ERROR("No points are available for pivot calibration");
     return IGSIO_FAIL;
```

Comment:
if it  is an error case then we should log an error; if not an error,  return success